### PR TITLE
Closes #88 Enable typescript on server

### PR DESCRIPTION
**Pull request by karlrez**
Closes https://github.com/mssiegel/frempco-client/issues/88

Installed TypeScript, types, and added a tsconfig. No changes to any existing code. 

tsconfig warning for 'no inputs' goes away once you add a TypeScript file to the project
image

### CODE REVIEW

**Review by zachlankton — CHANGES_REQUESTED** [review]
Good start, but we need to go a little further to setup typescript:

I've made some comments on the tsconfig.json that I think we should address.

We should add `/build` to our `.gitignore` file.

And we will need to make the following changes to our scripts in `package.json`

```json
 "scripts": {
    "dev": "nodemon -e ts --exec \"npm run dev:build\"",
    "start": "node build/index",
    "build": "tsc",
    "dev:build": "tsc && node build/index"
  },
```

We will need the `npm run build` step for Heroku to build the project before starting it with `npm start`.  @mssiegel is this something that you need to do in Heroku before we merge into main? (or dev if you get that setup)?

These script changes will point node to the build version and allow us to develop using the typescript compiler while still using nodemon to watch for changes and run the server.

Also, I think it safe to rename the following files to typescript (*.ts)

```bash
/index.js
/routes/classrooms.js
/services/corsOptions.js
/services/database.js
/services/socketIOSetup.js
```

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    "target": "es6",
+    "module": "commonjs",
+    "sourceMap": true,
+    "outDir": "./dist",
+    "strict": true,
+    "moduleResolution": "node",
+    "esModuleInterop": true,
+    "forceConsistentCasingInFileNames": true
+  }
+}
```
[the remark this answers is not in the document]
Comment (karlrez) [reply tsconfig.json:12]: Good point, It would be good to be consistent

**Comment by karlrez** [comment]
Thanks for the review @zachlankton, that was a very good explanation on the setup. I just made my updates as per your comments.

**Review by mssiegel — APPROVED** [review]
Great work addressing Zach's suggestions! I added new comments. Once they're addressed this can be merged!! Heroku should make it work automatically. 
I had suggestions for further typescript improvements but didn't want to block this PR so I made a [new issue](https://github.com/mssiegel/frempco-client/issues/89) for them. 

@karlrez To be extra safe, I'd like to merge this PR.  Once you address the comments, tag me and I'll go ahead and merge and then verify it works in production. 

Also, I am new to Typescript configurations. So feel free to argue back on anything I said!

File: package.json [hunk 2]
```diff
@@ -5,8 +5,10 @@
   "main": "index.js",
   "type": "module",
   "scripts": {
-    "dev": "nodemon index",
-    "start": "node index"
+    "dev": "nodemon -e ts --exec \"npm run dev:build\"",
+    "start": "node build/index",
+    "build": "tsc",
+    "dev:build": "tsc && node build/index"
```
Comment (mssiegel) [inline package.json:11]: `postinstall` tells Heroku what to do after installing the npm packages
```suggestion
    "dev:build": "tsc && node build/index",
    "postinstall": "tsc"
```

Also, add a new file to the root directory called Procfile and put in `web: node build/index.js`
This tells Heroku the entry point of our app as documented [here](https://medium.com/developer-rants/deploying-typescript-node-js-applications-to-heroku-81dd75424ce0)

**Review by mssiegel — CHANGES_REQUESTED** [review]
Whoops. I accidentally clicked the button to re-request a review from myself.

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    "lib": ["esnext"],
+    "allowJs": true,
+    "skipLibCheck": true,
+    "strict": false,
+    "forceConsistentCasingInFileNames": true,
+    "outDir": "./build",
+    "incremental": true,
+    "esModuleInterop": true,
+    "module": "esnext",
+    "moduleResolution": "node",
```
[the remark this answers is not in the document]
Comment (karlrez) [reply tsconfig.json]: After running `npm dev:build` i ran into this:

image

Updated TypeScript to this version, which resolved the issue:
image

Let me know if this works for you and I'll merge

**Review by mssiegel — APPROVED** [review]
Looks great! I'm going to merge!

**Review by mssiegel — CHANGES_REQUESTED** [review]
Don't merge yet. Just saw karls comment

File: tsconfig.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (mssiegel) [reply tsconfig.json]: @karlrez lets change it back to just `"node"`.  I don't know what the whole dev typescript version is and not worth spending the time to look into it.

[the remark this answers is not in the document]
Comment (karlrez) [reply tsconfig.json:13]: @mssiegel done, switched back to "node". 
The whole -dev typescript thing was auto generated when I updated, I dont think its required, it also works fine with just the version number

[the remark this answers is not in the document]
Comment (mssiegel) [reply tsconfig.json:13]: Thanks for changing it back. Anything that says 'X' is unstable isn't worth us using : )

**Review by mssiegel — APPROVED** [review]
Great work! Normally the person who makes the PR merges. But I'm going to merge this PR so that I can immediately verify it'll work on Heroku.

**Comment by mssiegel** [comment]
WHOOHOO!!! It works perfectly on production! GREAT TEAM EFFORT EVERYONE!! Getting the backend to use Typescript is a huge step forward!!!

**Pull request merged**